linera-storage/src/migration.rs

```diff
@@ -0,0 +1,572 @@
+            assert_eq!(schema, SchemaVersion::Version1);
+        } else {
+            // Starting from scratch, so write the Schema to avoid migrating
+            // later.
+            self.write_database_schema(&SchemaVersion::Version1).await?;
+        }
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::{
+        collections::{BTreeMap, HashMap},
+        marker::PhantomData,
+        ops::Deref,
+    };
+
+    use linera_base::{
+        crypto::CryptoHash,
+        identifiers::{
+            BlobId, BlobType, ChainId, EventId, GenericApplicationId, StreamId, StreamName,
+        },
+    };
+    #[cfg(feature = "rocksdb")]
+    use linera_views::rocks_db::RocksDbDatabase;
+    #[cfg(feature = "scylladb")]
+    use linera_views::scylla_db::ScyllaDbDatabase;
+    use linera_views::{
+        batch::Batch,
+        memory::MemoryDatabase,
+        random::make_deterministic_rng,
+        store::{
+            KeyValueDatabase, KeyValueStore, ReadableKeyValueStore, TestKeyValueDatabase,
+            WritableKeyValueStore,
+        },
+        ViewError,
+    };
+    use rand::Rng;
+    use test_case::test_case;
+
+    use crate::{
+        db_storage::RestrictedEventId,
+        migration::{BaseKey, RootKey, DEFAULT_KEY, ONE_KEY},
+        DbStorage, WallClock,
+    };
+
+    #[derive(Clone, Debug, Eq, PartialEq)]
+    #[allow(clippy::type_complexity)]
+    struct StorageState {
+        chain_ids_key_values: BTreeMap<ChainId, Vec<(Vec<u8>, Vec<u8>)>>,
+        certificates: BTreeMap<CryptoHash, Vec<u8>>,
+        confirmed_blocks: BTreeMap<CryptoHash, Vec<u8>>,
+        blobs: BTreeMap<BlobId, Vec<u8>>,
+        blob_states: BTreeMap<BlobId, Vec<u8>>,
+        events: HashMap<EventId, Vec<u8>>,
+        block_exporter_states: BTreeMap<u32, Vec<(Vec<u8>, Vec<u8>)>>,
+        network_description: Option<Vec<u8>>,
+    }
+
+    fn get_vector(rng: &mut impl Rng, len: usize) -> Vec<u8> {
+        let mut v = Vec::new();
+        for _ in 0..len {
+            let value = rng.gen::<u8>();
+            v.push(value);
+        }
+        v
+    }
+
+    fn get_hash(rng: &mut impl Rng) -> CryptoHash {
+        let rnd_val = rng.gen::<usize>();
+        CryptoHash::test_hash(format!("rnd_val={rnd_val}"))
+    }
+
+    fn get_stream_id(rng: &mut impl Rng) -> StreamId {
+        let application_id = GenericApplicationId::System;
+        let stream_name = StreamName(get_vector(rng, 10));
+        StreamId {
+            application_id,
+            stream_name,
+        }
+    }
+
+    fn get_event_id(rng: &mut impl Rng) -> EventId {
+        let hash = get_hash(rng);
+        let chain_id = ChainId(hash);
+        let stream_id = get_stream_id(rng);
+        let index = rng.gen::<u32>();
+        EventId {
+            chain_id,
+            stream_id,
+            index,
+        }
+    }
+
+    fn reorder_key_values(key_values: Vec<(Vec<u8>, Vec<u8>)>) -> Vec<(Vec<u8>, Vec<u8>)> {
```

Comment:
Why not just use `key_values.sort_unstable()`?

Reply:
Thanks. I wanted to have the test pass yesterday, and writing the code was faster than searching for it.